### PR TITLE
[SPARK-31864][SQL] Adjust AQE skew join trigger condition

### PR DESCRIPTION
**Pull request by maryannxue**
### What changes were proposed in this pull request?

This PR makes a minor change in deciding whether a partition is skewed by comparing the partition size to the median size of coalesced partitions instead of median size of raw partitions before coalescing.

### Why are the changes needed?

This change is line with target size criteria of splitting skew join partitions and can also cope with situations of extra empty partitions caused by over-partitioning. This PR has also improved skew join tests in AQE tests.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Updated UTs.

### CODE REVIEW

**Comment by maryannxue** [comment]
cc @cloud-fan @Ngone51

**Comment by cloud-fan** [comment]
This makes sense to me. e.g. for partitions [1, 1, 1, ..., 1, 10], which are coalesced to `[9, 10]`, I don't think there are skewed partitions.

@JkSelf what do you think?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 1]
```diff
@@ -164,15 +164,15 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
       assert(left.partitionsWithSizes.length == right.partitionsWithSizes.length)
       val numPartitions = left.partitionsWithSizes.length
       // We use the median size of the original shuffle partitions to detect skewed partitions.
```
Comment (JkSelf) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:166]: We may also need to adjust the comment here.

**Comment by JkSelf** [comment]
Good improvement. Except one small comment.  LGTM.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by manuzhang** [comment]
@cloud-fan @maryannxue @JkSelf I'm seeing a case where partitions [0,0,0,...,13GB] were coalesced to [13GB]  and took 17 min for a SortMergeJoin. With coalescing disabled, partitions would be split into [0,0,0,..., 256MB, 256MB,...,256MB] by OptimizeSkewedJoin and only took 38s. WDYT ?

**Comment by cloud-fan** [comment]
I think a single partition should be a special case. The median size doesn't make sense anymore and we should use the target partition size. Can you open a PR for it?

**Comment by maryannxue** [comment]
I assume there's one only key in that 13G partition. You can always run into this when the number of distinct keys is smaller than the partition number. It is not a good idea to use median whether it's pre-coalescing size or post-coalescing size. Suppose, you have 100 keys, and the partition number is set to 200, even with pre-coalescing size, you won't get the skew handling you want, but with partition number being 201 you can. The skew behavior now depends on the partition number. It's probably not a bad idea given we don't have a more advanced strategy here.
With something like 13G, we should probably always split it.

So think we can:
1) revert this PR, so for extremely clustered keys, skew can work
2) change the `spark.sql.adaptive.skewJoin.skewedPartitionFactor` check condition to ">= 0" instead of ">0" so we can force split large partitions even if there isn't an uneven distribution.

**Comment by manuzhang** [comment]
@cloud-fan 
I think there are two cases here. The first is only one big partition before coalescing. I don't think it fits into skew join optimization but more general splitting-big-partition optimization. The second is, like in my example, skew before coalescing but no skew afterwards. 

Do you want to target both cases or just the second ? 

@maryannxue 
In terms of performance, I don't think there will be a big difference between the case of "not skew join" 200 and optimizing "little skew join" 201. However, as in my case, the difference of optimizing "very skew join" before coalescing and "not skew join" after coalescing would be significant. 

I prefer option 1 since we can have a more general configuration that works good enough.

**Comment by cloud-fan** [comment]
I think the current status for 3.0 is good enough, as it's conservative to trigger the skew join optimization.

@manuzhang can you send a PR to master branch to do the revert? We can still keep the test change though.

**Comment by manuzhang** [comment]
@cloud-fan https://github.com/apache/spark/pull/28770 has been created. Please help review.

**Comment by maryannxue** [comment]
@manuzhang 
I was actually suggesting both :)

**Comment by manuzhang** [comment]
@maryannxue `spark.sql.adaptive.skewJoin.skewedPartitionFactor=0` doesn't look meaningful to me. Even `spark.sql.adaptive.skewJoin.skewedPartitionFactor=1` feels counter-intuitive as we are optimizing skew joins when there aren't any skewness.

**Comment by JkSelf** [comment]
@manuzhang If you adjust the configuration of `spark.sql.adaptive.skewJoin.skewedPartitionFactor ` between  0 and 1, can the 13GB partition be split? If so, can we set the skewed related configurations to trigger the skew optimization not revert this PR? I think this PR is meaningful when the rule of `OptimizeSkewedJoin` is behind `CoalesceShufflePartitions` in the `queryStageOptimizerRules` of `AdaptiveSparkPlanExec`.

**Comment by manuzhang** [comment]
@JkSelf That will work but the value doesn't make sense to me. In addition, I don't want to tune the configuration for every specific case but come up with a suite of default configurations that work out of box for most cases.

> I think this PR is meaningful when the rule of OptimizeSkewedJoin is behind CoalesceShufflePartitions in the queryStageOptimizerRules of AdaptiveSparkPlanExec

I don't see too much "meaning" here as we split the skewed partition towards the average size of coalesced non-skew partitions or advisory target size whichever is larger. We'll end up with evenly distributed partitions with or without this PR. However, we lose the chance to optimize skew partitions in some cases with this change.

**Pull request closed without merging**